packages/dynamite/dynamite_runtime/lib/src/http_client.dart

```diff
@@ -0,0 +1,129 @@
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:cookie_jar/cookie_jar.dart';
+import 'package:universal_io/io.dart';
+
+export 'package:cookie_jar/cookie_jar.dart';
+
+extension DynamiteHttpClientResponseBody on HttpClientResponse {
+  Future<Uint8List> get bodyBytes async {
+    final data = await expand((final element) => element).toList();
+
+    return Uint8List.fromList(data);
+  }
+
+  Future<String> get body => transform(utf8.decoder).join();
+
+  Future<dynamic> get jsonBody => transform(utf8.decoder).transform(json.decoder).first;
+
+  Map<String, String> get responseHeaders {
+    final responseHeaders = <String, String>{};
+    headers.forEach((final name, final values) {
+      responseHeaders[name] = values.last;
+    });
+
+    return responseHeaders;
+  }
+}
+
+class DynamiteResponse<T, U> {
+  DynamiteResponse(
+    this.data,
+    this.headers,
+  );
+
+  final T data;
+
+  final U headers;
+
+  @override
+  String toString() => 'DynamiteResponse(data: $data, headers: $headers)';
+}
+
+class DynamiteApiException implements Exception {
+  DynamiteApiException(
+    this.statusCode,
+    this.headers,
+    this.body,
+  );
+
+  final int statusCode;
+
+  final Map<String, String> headers;
+
+  final String body;
```

Comment:
Shouldn't this be Uint8List? Just in case we get some binary data but are unable to handle it for whatever reason and then throw the exception. On the other hand this would make it quite annoying to read the error message

Reply:
I could switch it back to the binary stuff and only utf.decode it when needed (performance isn't really a problem here) but the old way we just did the same.
Decode it in the toString method. Now if we have binary data the old way would probably have thrown an exception when calling toString. The new way at least throws this in the request itself where we catch various errors.

I think the best way would be to pass the binary data into this but still wrap the toString in a try catch and just printing `binary` in that case.

Reply:
I like that idea